sway-core/src/parse_tree/declaration/variable.rs

```diff
@@ -10,3 +18,231 @@ pub struct VariableDeclaration {
     pub body: Expression, // will be codeblock variant
     pub is_mutable: bool,
 }
+
+pub enum VariableDeclarationLHS {
+    Name(LHSName),
+    Tuple(LHSTuple),
+}
+
+pub struct LHSName {
+    name: Ident,
+    is_mutable: bool,
+    span: Span,
+}
+
+pub struct LHSTuple {
+    elems: Vec<VariableDeclarationLHS>,
+    span: Span,
+}
+
+impl VariableDeclaration {
+    pub(crate) fn parse_from_pair(
+        pair: Pair<Rule>,
+        config: Option<&BuildConfig>,
+    ) -> CompileResult<Vec<Self>> {
+        let mut warnings = vec![];
+        let mut errors = vec![];
+        let mut var_decl_parts = pair.into_inner();
+        let _let_keyword = var_decl_parts.next();
+        let lhs = check!(
+            VariableDeclarationLHS::parse_from_pair(
+                var_decl_parts.next().expect("gaurenteed by grammar"),
+                config
+            ),
+            return err(warnings, errors),
+            warnings,
+            errors
+        );
+        let mut maybe_body = var_decl_parts.next().unwrap();
+        let type_ascription = match maybe_body.as_rule() {
+            Rule::type_ascription => {
+                let type_asc = maybe_body.clone();
+                maybe_body = var_decl_parts.next().unwrap();
+                Some(type_asc)
+            }
+            _ => None,
+        };
+        let type_ascription_span = type_ascription.clone().map(|x| Span {
+            span: x.into_inner().next().unwrap().as_span(),
+            path: config.map(|x| x.path()),
+        });
+        let type_ascription = if let Some(ascription) = type_ascription {
```

Comment:
I think we should have positive/negative test cases for the following, would you mind adding them?
```rust
// 1
let (x, y): (u32, bool) = (10, true);
// 2
let (x, y): (u32, _) = (42, true);
// 3
fn test<T, E>(a: T, b: E) {
     let (x, y): (T, E) = (a, b);
} 
test(true, false);
// ensure monomorphizing works in tuples by instantiating with different types
test (42, 42);

// 4
let (a, (b, (c, d) ) ): (u64, (u32, (bool, str[2]) ) ) = (42u64, (42u32, (true, "ok") ) );

````